[a2av] Indexed push

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #168047
* #167792
* #167676
* #167144
* #166896
* #166899
* #166898
* #166774
* #163815
* #163814
* #166772
* __->__ #165482
* #165101
* #164181
* #164164

Based on `topk_indices`, push token to corresponding expert(s), once a token becomes ready in the input buffer, as indicated by a head pointer. 

This kernel is assumed to run concurrently with a "producer" kernel, e.g. the rail-wise all-to-all, while this kernel serves the intra-node all-to-all.

Here "expert" is a logical entity on a peer rank, and one rank may host multiple experts.

```
_all_to_all_v_2d_index_push(
  Tensor input, Tensor(a!) out,
  Tensor topk_indices, Tensor occurrences, Tensor dst_offsets, str group_name, 
  Tensor b_start, Tensor b_len, Tensor b_head,  # signals for each CUDA block
)
```

Status:
Test shows that the numerics is the same as the non-indexed version based on input expansion, i.e. `all_to_all_vdev_2d`.

Overall progress looks like this (because producer also has multiple blocks):
```
[█████░░░░][███░░░][██░░░░░░]
 Block 0   Block 1  Block 2
```
cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

kwen2501 committed:
Update

[ghstack-poisoned]

ngimel commented on torch/csrc/distributed/c10d/symm_mem/nvshmem_extension.cu:
```diff
@@ -1151,6 +1151,133 @@ void _make_a2a_2d_exchange_plan(
+      &topk, &n_local_experts, &stride_bytes, &team,
+      &b_start_ptr, &b_len_ptr, &b_head_ptr
+  };
+  C10_CUDA_CHECK(cudaLaunchKernel(
```
any reason to use `cudaLaunchKernel` instead of `<<<>>>`

kwen2501 committed:
Update on "[a2av] Indexed push"


Based on `topk_indices`, push token to corresponding expert(s), once a token becomes ready in the input buffer, as indicated by a head pointer. 

This kernel is assumed to run concurrently with a "producer" kernel, e.g. the rail-wise all-to-all, while this kernel serves the intra-node all-to-all.

Here "expert" is a logical entity on a peer rank, and one rank may host multiple experts.

```
_all_to_all_v_2d_index_push(
  Tensor input, Tensor(a!) out,
  Tensor topk_indices, Tensor occurrences, Tensor dst_offsets, str group_name, 
  Tensor b_start, Tensor b_len, Tensor b_head,  # signals for each CUDA block
)
```

Status:
Test shows that the numerics is the same as the non-indexed version based on input expansion, i.e. `all_to_all_vdev_2d`.

Overall progress looks like this (because producer also has multiple blocks):
```
[█████░░░░][███░░░][██░░░░░░]
 Block 0   Block 1  Block 2
```
cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]
